### PR TITLE
begin separation of constraints and modifiers

### PR DESCRIPTION
**Pull request by lukasheinrich**
# Description

This PR will separate modifier and contraint codes since some modifiers can share the same constraint (e.g. the qualify_names issue)

# Checklist Before Requesting Reviewer

- [x] Tests are passing
- [x] "WIP" removed from the title of the pull request
- [x] Selected an Assignee for the PR to be responsible for the log summary

# Before Merging

For the PR Assignees:

- [x] Summarize commit messages into a comprehensive review of the PR

### CODE REVIEW

**Comment by kratsg** [comment]
I think I understand what's going on here -- since you're separating out the initialization/alphas of the constraints themselves from the modifiers. So each modifier has a constraint object that it refers to. Rather than duping code across modifiers?

A naive question is wondering why not go more OOP and creating different types of constraints that modifiers inherit from instead of setting it as the constraint object on the modifier itself? E.G. `class histosys(standard_gaussian_constraint)`

**Comment by kratsg** [comment]
> @kratsg we do composition instead of inheritance precisely to address the qualify_names issue. Multiple modifier instances will point to the same constraint instance. Inheritance would keep the 1<->1 mapping

Only if you don't do meta-classing there to register things. But how do you ensure the same constraint instance?

**Comment by kratsg** [comment]
I still don't see how you have a 1<->1 mapping since histosys would never have the poisson constraint, for example - based on the way this is refactored.

**Comment by lukasheinrich** [comment]
this is not done yet. Right now each modifier does create its own instance (keeping 1<-> 1) .. in the end, we will inject the constraint object into the modifier on creation-time (where there will be some checks to assure that the constraint object is an instance of the constraint type that the modifier expects and that the dimensions are correct)

**Comment by kratsg** [comment]
> this is not done yet. Right now each modifier does create its own instance (keeping 1<-> 1) .. in the end, we will inject the constraint object into the modifier on creation-time (where there will be some checks to assure that the constraint object is an instance of the constraint type that the modifier expects and that the dimensions are correct)

I mean, it's still much easier, I think, to handle this with singletons instead to some extent, rather than trying to inject the function into the code from outside of the modifier code itself. Otherwise, this will start to get messy.

File: pyhf/constraints.py [hunk 1]
```diff
@@ -1,5 +1,36 @@
 from . import get_backend, default_backend
 
+
+class standard_gaussian_constraint(object):
```
Comment (kratsg) [inline pyhf/constraints.py]: call this "gaussian_constraint"

File: pyhf/constraints.py [hunk 2]
```diff
@@ -1,5 +1,36 @@
...
+    def __init__(self, n_parameters, inits, bounds, auxdata):
+        self.pdf_type = 'normal'
```
Comment (kratsg) [inline pyhf/constraints.py]: do we want to change `pdf_type` from `normal` to `gaussian`?

File: pyhf/constraints.py [hunk 3]
```diff
@@ -1,5 +1,36 @@
...
+        self.n_parameters = n_parameters
+        self.suggested_init = inits
+        self.suggested_bounds = bounds
+        self.auxdata = auxdata
+
+    def alphas(self, pars):
+        return pars  # the nuisance parameters correspond directly to the alpha
```
Comment (kratsg) [inline pyhf/constraints.py]: change this to the function docstring

File: pyhf/constraints.py [hunk 4]
```diff
@@ -1,5 +1,36 @@
...
+
+    def expected_data(self, pars):
+        return self.alphas(pars)
+
+class factor_poisson_constraint(object):
```
Comment (kratsg) [inline pyhf/constraints.py]: call this `poisson_constraint`

File: pyhf/constraints.py [hunk 5]
```diff
@@ -1,5 +1,36 @@
...
+    def __init__(self, n_parameters, inits, bounds, auxdata, factors):
+        self.factors = factors
+        self.pdf_type = 'poisson'
```
Comment (kratsg) [inline pyhf/constraints.py]: self.pdf_type should be first (to mirror `gaussian_constraint`)

File: pyhf/constraints.py [hunk 6]
```diff
@@ -1,5 +1,36 @@
...
+        self.n_parameters = n_parameters
+        self.suggested_init = inits
+        self.suggested_bounds = bounds
+        self.auxdata = auxdata
+
+    def alphas(self, pars):
+        tensorlib, _ = get_backend()
+        return tensorlib.product(tensorlib.stack([pars, tensorlib.astensor(self.factors)]), axis=0)
+    def expected_data(self, pars):
```
Comment (kratsg) [inline pyhf/constraints.py]: missing a newline before this

File: pyhf/constraints.py [hunk 7]
```diff
@@ -1,5 +1,36 @@
...
+class factor_poisson_constraint(object):
+    def __init__(self, n_parameters, inits, bounds, auxdata, factors):
+        self.factors = factors
+        self.pdf_type = 'poisson'
+        self.n_parameters = n_parameters
+        self.suggested_init = inits
+        self.suggested_bounds = bounds
+        self.auxdata = auxdata
+
+    def alphas(self, pars):
```
Comment (kratsg) [inline pyhf/constraints.py]: need a comment on why you're doing the product/stack thing (maybe explain the math here)

File: pyhf/paramsets.py [hunk 8]
```diff
@@ -0,0 +1,38 @@
+from . import get_backend
+class param_set(object):
```
Comment (kratsg) [inline pyhf/paramsets.py]: call this `paramset`

File: pyhf/paramsets.py [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    def __init__(self, n_parameters, inits, bounds):
+        self.n_parameters = n_parameters
+        self.suggested_init = inits
+        self.suggested_bounds = bounds
+
+class normal_constrained_set(param_set):
```
Comment (kratsg) [inline pyhf/paramsets.py]: call this `constrained_normal_paramset`

File: pyhf/paramsets.py [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+        self.auxdata = auxdata
+
+    def alphas(self, pars):
+        '''the nuisance parameters correspond directly to the alpha'''
+        return pars  
+
+    def expected_data(self, pars):
+        return self.alphas(pars)
+
+class poisson_constrained_set(param_set):
```
Comment (kratsg) [inline pyhf/paramsets.py]: call this `constrained_poisson_paramset`

File: pyhf/paramsets.py [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+        tensorlib, _ = get_backend()
+        return tensorlib.product(
+                tensorlib.stack([pars, tensorlib.astensor(self.factors)]
+            ),
+            axis=0
+        )
+
+    def expected_data(self, pars):
+        return self.alphas(pars)
+
```
Comment (kratsg) [inline pyhf/paramsets.py]: add a `class unconstrained_paramset(paramset)` and use that instead of the `paramset` class to help with naming.

**Pull request merged**